### PR TITLE
feat: add login notification for custom webhook handlers

### LINKED ISSUE #320: Feature request: Make generic achievement metadata more accesible
I am currently working on making an OSRS 2011-style adventurer's log (see pic). It is using Dink as it's engine for pushing log data (admittedly not the intended usecase). I chose Dink because it is avaiable in both HDOS as well as Runelite.

firefox_F7NBRXQF7b

repo: https://github.com/J3sven/OSRS-alog

One issue I am facing with Dink right now is that the (combat) achievement, Quest point and Collectionlog counters do no update until a relevant event has occured. A method of accessing this data more readily would be much appreciated, especially considering accounts with any of these stats maxed would never see an update until their cap is raised.

Some suggestions:

- Simply include the data in every webhook request (perhaps a bit overkill)
-  A generic log in/log out notification that lists the user's stats and achievement data. This could be something desirable for general users too as a reference of what progress they made that day.

### PR DESCRIPTION
**Pull request by iProdigy**
Closes #320 

Only notifies upon login (can be expanded in future PRs)

### CODE REVIEW

**Comment by Felanbird** [comment]
> should there be a webhook override box for this notifier? 

~~There probably has to be only because users who use overrides likely wouldn't have anything in the main box, and directing them to input someones URL into the Primary box to use this specific feature could cause confusion. (using overrides for everything else)~~

In a perfect world this has a box in advanced that only does this job, unrelated to any other webhooks set.


> should the notifier enabled box be moved to a separate section than advanced?

seems fine in advance, one day we'll probably need to expand `Advanced` to `General` & `Advanced` so stuff like `hide chat` aren't hidden in the spooky settings

> should we allow sending this notification to discord (e.g., with a customizable cooldown)?

the data is better used in a UI versus a discord notification, even with a configurable delay it seems like a waste to send discord notifications saying hey I logged in and my questpoints & combatpoints are probably the same. Obviously works better with custom handling where they can throw notifs in the garbage where applicable

**Comment by pajlada** [comment]
>     * should there be a webhook override box for this notifier?

I think not, no. I think it might be easier if this is just a textbox where you insert a custom URL  

>     * should the notifier enabled box be moved to a separate section than advanced?

I think, with the change above, no enabled checkbox should be necessary. Just set & forget an advanced webhook URL.

>     * should we allow sending this notification to discord (e.g., with a customizable cooldown)?

I don't think so


Maybe renaming this to a "meta notifier" or "metadata notifier" would make sense, which would allow us to just send random events to this specific webhook, with the only webhook type right now being the "login" webhook?

**Review by Felanbird — APPROVED** [review]
```{
   "type":"LOGIN",
   "playerName":"felanbird1",
   "accountType":"NORMAL",
   "clanName":"P i c k l e",
   "extra":{
      "world":494,
      "combatAchievementPoints":{
         "completed":503,
         "total":2005
      },
      "achievementDiary":{
         "completed":42,
         "total":48
      },
      "barbarianAssault":{
         "highGambleCount":0
      },
      "skills":{
         "totalExperience":337778499,
         "totalLevel":2164,
         "levels":{
            "Hunter":90,
            "Thieving":86,
            "Runecraft":86,
            "Construction":86,
            "Cooking":103,
            "Magic":106,
            "Fletching":99,
            "Herblore":91,
            "Firemaking":100,
            "Attack":107,
            "Fishing":92,
            "Crafting":96,
            "Hitpoints":111,
            "Ranged":110,
            "Mining":88,
            "Smithing":91,
            "Agility":82,
            "Woodcutting":96,
            "Slayer":104,
            "Defence":103,
            "Strength":104,
            "Prayer":91,
            "Farming":100
         }
      },
      "questCount":{
         "completed":156,
         "total":158
      },
      "questPoints":{
         "completed":296,
         "total":300
      },
      "slayer":{
         "points":2204,
         "streak":1074
      }
   },
}
```

Login after having the `Character summary` tab loaded on logout

```
{
   "type":"LOGIN",
   "playerName":"felanbird1",
   "accountType":"NORMAL",
   "clanName":"P i c k l e",
   "extra":{
      "world":494,
      "collectionLog":{
         "completed":651,
         "total":1477
      },
```

looks like this if the user was to input a discord webhook
image

**Review by pajlada — APPROVED** [review]

**Comment by pajlada** [comment]
@j3sven could you take a peek at the payload posted above to see if it looks as expected for you?

**Comment by J3sven** [comment]
> @J3sven could you take a peek at the payload posted above to see if it looks as expected for you?

```
      "achievementDiary":{
         "completed":42,
         "total":48
      },
```
Could this maybe include total tasks completed as well? rather than diary sets? That's what is displayed under character summary ingame so that's what I modeled the counter on the alog profile after. I noticed that is not included in the usual diary payload either so if it is a big ask then no worries.

I would also like to include the collection log slots/total to be included. 

For some clarity from my side, the elements I am trying to cover with this data are these:
image


Apologies if I was not clear enough about the specifications in my original issue. Aside from these two point it looks great though!

**Comment by iProdigy** [comment]
> I would also like to include the collection log slots/total to be included.

This *is* included, but only of the client knows the data (i.e., character summary tab is selected, otherwise jagex doesn't send the varbit on login)


> Could this maybe include total tasks completed as well? rather than diary sets?

We will investigate (there may be a similar data availability issue)

**Comment by J3sven** [comment]
> > I would also like to include the collection log slots/total to be included.
> 
> This _is_ included, but only of the client knows the data (i.e., character summary tab is selected, otherwise jagex doesn't send the varbit on login)

Ahh, that makes sense. Oh well good that it's included when available though, I can always have a setup process that lets people relog with the tab open or something.

**Comment by Felanbird** [comment]
```
{
   "type":"LOGIN",
   "playerName":"felanbird1",
   "accountType":"NORMAL",
   "clanName":"P i c k l e",
   "extra":{
      "world":338,
      "collectionLog":{
         "completed":651,
         "total":1477
      },
      "combatAchievementPoints":{
         "completed":503,
         "total":2005
      },
      "achievementDiary":{
         "completed":42,
         "total":48
      },
      "achievementDiaryTasks":{
         "completed":477,
         "total":492
      },
      "barbarianAssault":{
         "highGambleCount":0
      },
      "skills":{
         "totalExperience":337810454,
         "totalLevel":2164,
         "levels":{
            "Hunter":90,
            "Thieving":86,
            "Runecraft":86,
            "Construction":86,
            "Cooking":103,
            "Magic":106,
            "Fletching":99,
            "Herblore":91,
            "Firemaking":100,
            "Attack":107,
            "Fishing":92,
            "Crafting":96,
            "Hitpoints":111,
            "Ranged":110,
            "Mining":88,
            "Smithing":91,
            "Agility":82,
            "Woodcutting":96,
            "Slayer":104,
            "Defence":103,
            "Strength":104,
            "Prayer":91,
            "Farming":100
         }
      },
      "questCount":{
         "completed":156,
         "total":158
      },
      "questPoints":{
         "completed":296,
         "total":300
      },
      "slayer":{
         "points":2204,
         "streak":1074
      }
   },
}
```

File: README.md [hunk 1]
```diff
@@ -746,6 +747,87 @@ See [javadocs](https://static.runelite.net/api/runelite-api/net/runelite/api/Gra
...
 </details>
 
+### Metadata:
+
+On login, Dink can submit a character summary containing data that spans multiple notifiers to a custom webhook handler (configurable in the `Advanced` section).
+
+<details>
+  <summary>JSON for Login Notifications:</summary>
+
+```json5
```
Comment (Mm2PL) [inline README.md:757]: Please include information about `collectionLog` not being always sent. This also sent `clanName` for me which is missing from here. I'd like for this to include that we delay this webhook a little from the actual login.

**Review by Mm2PL — CHANGES_REQUESTED** [review]
just documentation nitpicks

**Review by Mm2PL — APPROVED** [review]

**Pull request merged**